### DISCUSSION TITLE
推荐一个 Go 项目发布工具

### DISCUSSION THREAD

**Discussion by sixg0000d** [IrineSistiana/mosdns#134] [General]
前几天在摸索 Github CLI 是如何打包的，发现他们用的是这个东西：
[GoReleaser](https://goreleaser.com/) 以及其配套 [GitHub Actions](https://github.com/goreleaser/goreleaser-action) ，除了 GitHub Actions 还能配合很多 CI 使用，能够代替自己写的脚本进行编译打包并上传的工作，简单且配置灵活，除了打包普通压缩包甚至还能打包 deb 、 rpm 等常见 Linux 包管理器包，感觉非常适合 mosdns 这样的项目。

我在自己的 fork 里按 release.py 的行为写了个行为相仿的 goreleaser 配置文件并试了试打包 deb 和 rpm ，[效果在这](https://github.com/sixg0000d/mosdns/releases)。

虽然有几个小缺陷：
- 不能改变文件打包进压缩包的路径 https://github.com/goreleaser/goreleaser/issues/1638
- 环境变量不能用命令导入，所以 VERSION 得在工作流里完成。不过说实话他内置的 `.Version` 的实现也挺不错的，所以推荐直接用这个而不用想办法去 `VERSION=$(git describe --tags --long --always)` 。
- 使用 upx 不如你的 release.py 灵活。且由于 freebsd-amd64 和 mips64-hardfloat 的二进制文件不能 upx 压缩，直接在 hook 里使用 upx 会报错然后失败，还得额外用个脚本过渡下。

```bash
#!/bin/sh

upx -q -9 "$1" >/dev/null || :

```

**Comment by IrineSistiana** [comment]
非常方便. 之前我也在 mosdns-cn 上试了试.

>除了打包普通压缩包甚至还能打包 deb 、 rpm 等常见 Linux 包管理器包，

没注意还有这些功能,有空我去试试 

(๑•̀ㅂ•́)و✧
